hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/spark/sql/execution/datasources/HoodieMultipleBaseFileFormat.scala

```diff
@@ -0,0 +1,278 @@
+                                              options: Map[String, String],
+                                              hadoopConf: Configuration): PartitionedFile => Iterator[InternalRow] = {
+    val outputSchema = StructType(requiredSchema.fields ++ partitionSchema.fields)
+    val requiredSchemaWithMandatory = if (!isMOR || MergeOnReadSnapshotRelation.isProjectionCompatible(tableState.value)) {
+      // add mandatory fields to required schema
+      val added: mutable.Buffer[StructField] = mutable.Buffer[StructField]()
+      for (field <- mandatoryFields) {
+        if (requiredSchema.getFieldIndex(field).isEmpty) {
+          val fieldToAdd = dataSchema.fields(dataSchema.getFieldIndex(field).get)
+          added.append(fieldToAdd)
+        }
+      }
+      val addedFields = StructType(added.toArray)
+      StructType(requiredSchema.toArray ++ addedFields.fields)
+    } else {
+      dataSchema
+    }
+
+    val (parquetBaseFileReader, orcBaseFileReader, preMergeParquetBaseFileReader, preMergeOrcBaseFileReader) = buildFileReaders(
+      sparkSession, dataSchema, partitionSchema, requiredSchema, filters, options, hadoopConf, requiredSchemaWithMandatory)
+
+    val broadcastedHadoopConf = sparkSession.sparkContext.broadcast(new SerializableConfiguration(hadoopConf))
+    (file: PartitionedFile) => {
+      val filePath = sparkAdapter.getSparkPartitionedFileUtils.getPathFromPartitionedFile(file)
+      val fileFormat = detectFileFormat(filePath.toString)
+      file.partitionValues match {
+        case fileSliceMapping: PartitionFileSliceMapping =>
+          if (FSUtils.isLogFile(filePath)) {
+            // no base file
+            val fileSlice = fileSliceMapping.getSlice(FSUtils.getFileId(filePath.getName).substring(1)).get
+            val logFiles = getLogFilesFromSlice(fileSlice)
+            val outputAvroSchema = HoodieBaseRelation.convertToAvroSchema(outputSchema, tableName)
+            new LogFileIterator(logFiles, filePath.getParent, tableSchema.value, outputSchema, outputAvroSchema,
+              tableState.value, broadcastedHadoopConf.value.value)
+          } else {
+            // We do not broadcast the slice if it has no log files
+            fileSliceMapping.getSlice(FSUtils.getFileId(filePath.getName)) match {
+              case Some(fileSlice) =>
+                val hoodieBaseFile = fileSlice.getBaseFile.get()
+                val baseFileFormat = detectFileFormat(hoodieBaseFile.getFileName)
+                val partitionValues = fileSliceMapping.getInternalRow
+                val logFiles = getLogFilesFromSlice(fileSlice)
+                if (requiredSchemaWithMandatory.isEmpty) {
+                  val baseFile = createPartitionedFile(partitionValues, hoodieBaseFile.getHadoopPath, 0, hoodieBaseFile.getFileLen)
+                  baseFileFormat match {
+                    case "parquet" => parquetBaseFileReader(baseFile)
+                    case "orc" => orcBaseFileReader(baseFile)
```

Comment:
Maybe we can avoid to hardcode these file format constants.

Reply:
No, it will ultimately fallback to using the table config because the `HoodieSyncConfig.META_SYNC_BASE_FILE_FORMAT` has infer function - https://github.com/apache/hudi/blob/0ad4560f2a4de00e43814b0d6cef2886a8a38155/hudi-sync/hudi-sync-common/src/main/java/org/apache/hudi/sync/common/HoodieSyncConfig.java#L103

Reply:
Good point! We do have `HoodieFileFormat` enum which can be used here. Let me do that in a minor followup PR where I will refactor some parts. I have HUDI-6986 to track the refactoring.